apps/manager_app/auth_card.c

```diff
@@ -0,0 +1,542 @@
+
+  // Copy response to pb result struct
+  memcpy(&(result.auth_card), resp, sizeof(manager_auth_card_response_t));
+
+  // Encode result to byte-stream
+  if (false ==
+      encode_manager_result(&result, buffer, sizeof(buffer), &bytes_written)) {
+    // Encoding failed, return false
+    return MANAGER_TASK_ENCODING_FAILED;
+  }
+
+  usb_send_msg(buffer, bytes_written);
+  return MANAGER_TASK_SUCCESS;
+}
+
+static void prepare_card_auth_context(
+    manager_auth_card_initiate_request_t *initiate_request) {
+  ASSERT(NULL != initiate_request);
+
+  auth_card_ctx.acceptable_cards = ACCEPTABLE_CARDS_ALL;
+
+  if (initiate_request->has_card_index) {
+    auth_card_ctx.acceptable_cards =
+        encode_card_number(initiate_request->card_index);
+    snprintf(auth_card_ctx.heading,
+             sizeof(auth_card_ctx.heading),
+             UI_TEXT_TAP_CARD,
+             tap_card_data.acceptable_cards);
+  } else {
+    snprintf(auth_card_ctx.heading,
+             sizeof(auth_card_ctx.heading),
+             ui_text_tap_a_card);
+  }
+
+  auth_card_ctx.pair_card_required = false;
+  if (initiate_request->has_pair_card) {
+    auth_card_ctx.pair_card_required = initiate_request->pair_card;
+  }
+  snprintf(auth_card_ctx.message,
+           sizeof(auth_card_ctx.message),
+           UI_TEXT_PLACE_CARD_TILL_BEEP,
+           SIGN_SERIAL_BEEP_COUNT(auth_card_ctx.pair_card_required));
+}
+
+bool wait_for_user_confirmation() {
+  confirm_scr_init(ui_text_start_verification_of_card);
+
+  evt_status_t status = get_events(EVENT_CONFIG_UI, MAX_INACTIVITY_TIMEOUT);
+
+  if (status.ui_event.event_occured) {
+    return status.ui_event.event_type == UI_EVENT_CONFIRM;
+  } else {
+    HANDLE_P0_EVENTS();
+  }
+  return false;
+}
+
+manager_error_code_t handle_sign_card_serial(
+    manager_auth_card_initiate_request_t *initiate_request,
+    manager_auth_card_response_t *resp) {
+  if (NULL == initiate_request || NULL == resp) {
+    return MANAGER_TASK_INVALID_ARGS;
+  }
+
+  instruction_scr_init(auth_card_ctx.message, auth_card_ctx.heading);
+  nfc_en_select_card_task();
+
+  evt_status_t status = get_events(EVENT_CONFIG_NFC, MAX_INACTIVITY_TIMEOUT);
+
+  if (true == status.p0_event.flag) {
+    HANDLE_P0_EVENTS();
+    return MANAGER_TASK_P0_ABORT_OCCURED;
+  }
+
+  // NFC event occured
+
+  /* TODO: Add sign serial card action*/
+
+  /* TODO: Check if card operation failed*/
+  // return MANAGER_TASK_FAILED;
+
+  UPDATE_AUTH_CARD_FLOW_STATUS(MANAGER_AUTH_CARD_STATUS_SERIAL_SIGNED);
+
+  // Display text for challenge sign screen
+  snprintf(auth_card_ctx.message,
+           sizeof(auth_card_ctx.message),
+           UI_TEXT_PLACE_CARD_TILL_BEEP,
+           SIGN_CHALLENGE_BEEP_COUNT(auth_card_ctx.pair_card_required));
+  instruction_scr_init(auth_card_ctx.message, auth_card_ctx.heading);
+
+  resp->which_response = MANAGER_AUTH_CARD_RESPONSE_SERIAL_SIGNATURE_TAG;
+  return MANAGER_TASK_SUCCESS;
+}
+
+manager_error_code_t handle_sign_challenge(
```

Comment:
Hi Aman, 

Let's move card flows to different file.

This way it is easy to differentiate between USB request handler - and then NFC flows

Reply:
The intent of this function is to call the sign serial card handler, and set status, UI, etc in response to it's success or failure.

I would add a to-do to separate out the NFC event. But IMO the function should still remain here.